### PR TITLE
Install GPG for signing Git commits

### PR DESCRIPTION
**Pull request by croaky**
http://mikegerwitz.com/papers/git-horror-story

### CODE REVIEW

**Comment by croaky** [comment]
gpg2 also installs `gpg-agent`

**Comment by pbrisbin** [comment]
:+1:

I wonder, what is laptop's stance in cases like this where there is additional setup required by the user to fully use this thing laptop is installing?

So far, we either go ahead and do it for them (e.g. `append_to_zsh`) or we're installing something that's fine to use without further configuration (e.g. `tmux`). In this case, `gpg2` is is not usable (at least not for the goal stated in laptop's README) unless the user takes a number of additional, non-trivial steps. Is that OK?

**Comment by croaky** [comment]
My opinion is Laptop's primary responsibility is installing (or updating) the necessary system dependencies as quickly as possible.

If we can reliably configure it in a way that will be compatible with the user's dotfiles, or if something needs to be started/restarted (like a database server), we should do those things as a secondary concern.

In this case, what are some additional things we would need to do? Start gpg-agent?

**Comment by pbrisbin** [comment]
> If we can [...], we should do those things as a secondary concern

I agree, and was asking for clarity around:

- When should we or should we not do those things?

I'd agree with your reasons, but would add my own that the thing we need to do should not require user input and should a simple thing not requiring complex shell code to get right. If we can't reliably do it without user intervention or difficult-to-maintain shell code, we should leave it up to the user.

- What do we do when we *can't* do this extra configuration and it's required for the stated goal of the tool we're installing (as is this case)?

I think I'd lean towards installing it anyway, but I can see the argument that if the user has to take 10 steps to do thing X anyway, what is the value in laptop doing the `brew install x` step for them?

> In this case, what are some additional things we would need to do?

The user would need a PGP key. This is non-trivial and requires user input or laptop to make many assumptions that it shouldn't -- see [caleb's post](http://robots.thoughtbot.com/pgp-and-you).

Once they had a key, we'd need to configure git to use it. That'd be pretty easy and (IMO) well within the realm of laptop to do. Unfortunately, I don't think laptop should be doing the first thing.

File: README.md [hunk 1]
```diff
@@ -45,6 +45,7 @@ What it sets up
 * [Exuberant Ctags] for indexing files for vim tab completion
 * [Foreman] for managing web processes
 * [gh] for interacting with the GitHub API
+* [GPG] for signing Git commits
```
Comment (mike-burns) [inline README.md]: Its formal name is "GnuPG".

**Comment by croaky** [comment]
@sikachu @georgebrock @harryschwartz I'd be interested in your thoughts on this PR. Anything else Laptop should be doing?

**Comment by georgebrock** [comment]
I agree with @pbrisbin: installing GnuPG is fine, but laptop probably shouldn't try to do anything beyond that.

If we want to encourage all thoughtbot employees to sign their Git commits, maybe we should mention it in the guides?

(Also: @harryschwartz should probably have been @hrs)

**Comment by hrs** [comment]
I'm not a laptop user, so I'm not sure if this falls outside its scope, but after installing GnuPG I'll generally:

* symlink `gpg` to `gpg2`
* append `ask-cert-level` to the end of my `~/.gnupg/gpg.conf`, since otherwise I invariably forget to use the `--ask-cert-level` flag when signing a key.

I like the idea of advocating signing in the guides. It's [very easy to set up](https://github.com/git/git/commit/2af2ef3c85612946cac3f16143b4875c9f31c4d2) as of git 2.0.

**Comment by georgebrock** [comment]
> append `ask-cert-level` to the end of my `~/.gnupg/gpg.conf`, since otherwise I invariably forget to use the `--ask-cert-level` flag when signing a key.

That's definitely useful, but it seems more like a dotfiles thing than a laptop thing.

**Comment by calebthompson** [comment]
> append ask-cert-level to the end of my ~/.gnupg/gpg.conf, since otherwise I invariably forget to use the --ask-cert-level flag when signing a key.

Ooh, thank you, @hrs

**Comment by croaky** [comment]
> if the user has to take 10 steps to do thing X anyway, what is the value in laptop doing the brew install x step for them?

Closed without merging for this reason.

**Pull request closed without merging**